[DOC-3386] Add docs for rotating a CMEK key (#14453)
cockroachdb/docs: cockroachcloud/managing-cmek.md

- [Enabling CMEK](#enable-cmek) for a {{ site.data.products.dedicated }} requires several steps:

>   - Turn on the {{ site.data.products.dedicated }} CMEK feature.

Author: ```suggestion
  - Request enrollment of your {{ site.data.products.db }} organization in the CMEK preview.
```

  - Provisioning an encryption key with your cloud provider's key management service (KMS).

>   - Granting {{ site.data.products.db }} access to the CMEK i.e., the key you control.

Reviewer: ```suggestion
  - Granting {{ site.data.products.db }} access to the CMEK (i.e., the key you control).
```

>   - Switching your cluster to use the CMEK.

Reviewer: Is there a more precise phrase than "Switching your cluster"? I only have the highest-level understanding from this.

> - [Checking CMEK status](#check-cmek-status) allows you to inspect the CMEK state of your cluster with a call to the Cluster API.

Reviewer: ```suggestion
- [Check CMEK status](#check-cmek-status) of your cluster with a call to the Cluster API.
```

> - [Rotate a CMEK key](#rotate-a-cmek-key) allows you to begin using a new CMEK key for one or more cluster regions.

Reviewer: ```suggestion
- [Rotating a CMEK key](#rotate-a-cmek-key) to begin using a new CMEK key for one or more cluster regions.
```

1. To start the process, contact {{ site.data.products.dedicated }} by reaching out to your account team, or [creating a support ticket](https://support.cockroachlabs.com/). You must provide your **Organization ID**, which you can find in your [Organization settings page](https://cockroachlabs.cloud/settings).

>     They will enable the CMEK feature for your {{ site.data.products.dedicated }} Organization.

Author: ```suggestion
    They will enable the CMEK feature for your {{ site.data.products.db }} Organization.
```

1. [Create a {{ site.data.products.db }} service account](console-access-management.html#service-accounts).
1. [Create an API key](console-access-management.html#create-api-keys) for the service account to use.

### Step 2. Provision your cluster

Create a new {{ site.data.products.dedicated }} cluster. There are two ways to do this:

- [Using the {{ site.data.products.db }} console clusters page](https://cockroachlabs.cloud/cluster).
- [Using the Cloud API](cloud-api.html#create-a-new-cluster). 

### Step 3. Provision IAM and KMS in your Cloud

Next, you must provision the resources required resources in your Cloud, whether this is AWS or GCP:

1. The key itself.
1. The authorization object that grants the {{ site.data.products.db }} service account permission to encrypt and decrypt using the key. The authorization object is either an IAM role in AWS or a service account in GCP.

> Follow the instructions for your {{ site.data.products.db }} Information-As-A-Service (IAAS) platform:

Reviewer 2: Can I request to please keep it cloud provider, and not IaaS (across this whole CMEK doc)? I'm just not comfortable with the IaaS term all across our docs. AWS, GCP and Azure are not just IaaS providers, they've PaaS & SaaS services too. In fact, KMS is a SaaS service. I don't know who came up with this idea of referring to them as IaaS platforms, but this needs correction across our docs.

See the [API specification](../api/cloud/v1.html#operation/CockroachCloud_EnableCMEK).

> 1. Create a new file named `cmek_config.json`. This file will contain a JSON array of `region_spec` objects, each of which includes the name of a {{ site.data.products.db }} region and a `key_spec` that is specific to the target KMS platform and specifies the URI of the CMEK key and the credential that authorizes {{ site.data.products.db }} to encrypt and decrypt using the key.

Reviewer 2: We should probably change "credential" to "authorization object" as that's what's being mentioned above.

Reviewer 3: Ah yeah. Consider "principal that is authorized", too. We call it `auth_principal`

Author: Went with the second suggestion as it reads more cleanly.

    Start from the example for your KMS platform and replace the placeholder values. Each of these examples includes `region_spec` objects for two {{ site.data.products.db }} regions; when enabling CMEK, you must include a `region_spec` for each region in the cluster.
    - **AWS**:

        {% include_cached copy-clipboard.html %}
        ~~~ json
        {
          "region_specs": [
            {

>               "region": "{COCKROACH_CLOUD_REGION}",

Reviewer 2: Should refer as COCKROACHDB_CLOUD_REGION at all places.

              "key_spec": {
                "type": "AWS_KMS",

>                 "uri": "arn:aws:kms:{KEY_URI}",

Reviewer 2: I would suggest to use an overall placeholder as {YOUR_AWS_KMS_KEY_ARN}. Specific value before a placeholder can make it confusing, and it's currently missing the region and AWS account id

                "auth_principal": "arn:aws:iam::{IAM_ROLE_NAME}"
              }
            },
            {
              "region": "{COCKROACH_CLOUD_REGION}",
              "key_spec": {
                "type": "AWS_KMS",
                "uri": "arn:aws:kms:{KEY_URI}",
                "auth_principal": "arn:aws:iam::{IAM_ROLE_NAME}"
              }
            }
          ]
        }
        ~~~
    - **GCP**:

        {% include_cached copy-clipboard.html %}
        ~~~ json
        {
          "region_specs": [
            {
              "region": "{COCKROACH_CLOUD_REGION}",
              "key_spec": {
                "type": "GCP_CLOUD_KMS",

>                 "uri": "arn:aws:kms:{KEY_RESOURCE_LOCATION}",

Reviewer 2: This is wrong. Let's just have a overall placeholder as {YOUR_GCP_KMS_KEY_RESOURCE_ID_EXCLUDING_VERSION}

>                 "auth_principal": "{COCKROACH_CLOUD_SERVICE_ACCOUNT}"

Reviewer 2: This is wrong. Let's just have a overall placeholder as {YOUR_GCP_SERVICE_ACCOUNT_ID}

See the [API specification](../api/cloud/v1.html#operation/CockroachCloud_EnableCMEK).

> 1. Create a new file named `cmek_config.json`. This file will contain JSON array of `region_spec` objects, each of which includes the name of a {{ site.data.products.db }} region and a `key_spec` that is specific to the target KMS platform and specifies the URI of the CMEK key and the credential that authorizes {{ site.data.products.db }} to encrypt and decrypt using the key.

Reviewer: ```suggestion
1. Create a new file named `cmek_config.json`. This file will contain a JSON array of `region_spec` objects, each of which includes the name of a {{ site.data.products.db }} region and a `key_spec` that is specific to the target KMS platform and specifies the URI of the CMEK key and the credential that authorizes {{ site.data.products.db }} to encrypt and decrypt using the key.
```

                "uri": "arn:aws:kms:{KEY_URI}",
                "auth_principal": "arn:aws:iam::{IAM_ROLE_NAME}"
              }
            }
          ]
        }
        ~~~
    - **GCP**:

        {% include_cached copy-clipboard.html %}
        ~~~ json
        {
          "region_specs": [
            {
              "region": "{COCKROACH_CLOUD_REGION}",
              "key_spec": {
                "type": "GCP_CLOUD_KMS",
                "uri": "arn:aws:kms:{KEY_RESOURCE_LOCATION}",
                "auth_principal": "{COCKROACH_CLOUD_SERVICE_ACCOUNT}"
              }
            },
            {
              "region": "{COCKROACH_CLOUD_REGION}",
              "key_spec": {
                "type": "GCP_CLOUD_KMS",
                "uri": "{KEY_RESOURCE_LOCATION}",
                "auth_principal": "{COCKROACH_CLOUD_SERVICE_ACCOUNT}"
              }
            }
          ]
        }
        ~~~

    {{site.data.alerts.callout_info}}
    Keep the following points in mind for multi-region clusters:
    - When enabling CMEK on a multi-region cluster, your `region_spec` must include an entry for each of the cluster's regions. Otherwise, an error occurs and CMEK is not enabled.

>     - If you add a new region to a cluster, data at rest in the new region is not automatically protested by CMEK. To enable CMEK for the new region, refer to [Rotate a CMEK key](#rotate-a-cmek-key).

Author: LOL

  --header "Authorization: Bearer ${API_KEY}"
```

## Rotate a CMEK key

{% include cockroachcloud/cmek-rotation-types.md %}

> The API to rotate a CMEK key is nearly identical to the API to [activate CMEK on a cluster](#step-4-activate-cmek-for-your-cockroachdb-dedicated-cluster), with one notable exception. When you activate CMEK, you use a `POST` request that includes a CMEK key for each of the cluster's regions. When you rotate a CMEK key, you use a `PUT` request that includes a CMEK key for each region you intend to rotate.

Reviewer 2: ```suggestion
The API to rotate a CMEK key is nearly identical to the API to [activate CMEK on a cluster](#step-4-activate-cmek-for-your-cockroachdb-dedicated-cluster), with one notable exception. When you activate CMEK, you use a `POST` request that includes a CMEK key for each of the cluster's regions. When you rotate a CMEK key, you use a `PUT` request that includes a CMEK key for each region where you intend to rotate.

Note: Do not disable or delete the previous CMEK key until the rotation process is complete.
```

Reviewer 3: Should we mention that `POST` can also be used to rotate, as long as it has all regions? Or would that confuse the narrative?

Reviewer 2: I would prefer to skip that. It's already a lot of nuance to absorb :)

See the [API specification](../api/cloud/v1.html#operation/CockroachCloud_EnableCMEK). <!-- TODO update when available -->

> 1. Create a new file named `cmek_config.json`. This file will contain a JSON array of `region_spec` objects, each of which includes the name of a {{ site.data.products.db }} region and a `key_spec` that is specific to the target KMS platform and specifies the URI of the CMEK key and the credential that authorizes {{ site.data.products.db }} to encrypt and decrypt using the key.

Reviewer 2: ```suggestion
1. Create a new file named `cmek_config.json`. This file will contain a JSON array of `region_spec` objects, each of which includes the name of a {{ site.data.products.db }} region and a `key_spec` that is specific to the target KMS platform and specifies the URI of the CMEK key and the authorization object that authorizes {{ site.data.products.db }} to encrypt and decrypt using the key.
```

>     Start from the example for your KMS platform and replace the placeholder values. Each of these examples includes `region_spec` objects for two {{ site.data.products.db }} regions; you need only include regions you want to update.

Reviewer 2: ```suggestion
    Start from the example for your KMS platform and replace the placeholder values. Each of these examples includes `region_spec` objects for two {{ site.data.products.db }} regions; you need only include regions you want to rotate CMEK in.
```

    - **AWS**:

        {% include_cached copy-clipboard.html %}
        ```json
        {
        "region_specs": [
          {

>             "region": "{COCKROACH_CLOUD_REGION}",

Reviewer 2: ```suggestion
            "region": "{COCKROACHDB_CLOUD_REGION}",
```

            "key_spec": {
              "type": "AWS_KMS",

>               "uri": "arn:aws:kms:{KEY_URI}",

Reviewer 2: ```suggestion
              "uri": "{YOUR_AWS_KMS_KEY_ARN}",
```

>               "auth_principal": "arn:aws:iam::{IAM_ROLE_NAME}"

Reviewer 2: ```suggestion
              "auth_principal": "{YOUR_AWS_IAM_ROLE_ARN}"
```

            }
          },
          {

>             "region": "{COCKROACH_CLOUD_REGION}",

Reviewer 2: ```suggestion
            "region": "{COCKROACHDB_CLOUD_REGION}",
```

            "key_spec": {
              "type": "AWS_KMS",

>               "uri": "arn:aws:kms:{KEY_URI}",

Reviewer 2: ```suggestion
              "uri": "{YOUR_AWS_KMS_KEY_ARN}",
```

>               "auth_principal": "arn:aws:iam::{IAM_ROLE_NAME}"

Reviewer 2: ```suggestion
              "auth_principal": "{YOUR_AWS_IAM_ROLE_ARN}"
```

            }
          }
        ]
      }
      ```
    - **GCP**:

        {% include_cached copy-clipboard.html %}
        ```json
        {
        "region_specs": [
          {

>             "region": "{COCKROACH_CLOUD_REGION}",

Reviewer 2: ```suggestion
            "region": "{COCKROACHDB_CLOUD_REGION}",
```

            "key_spec": {
              "type": "GCP_CLOUD_KMS",

>               "uri": "arn:aws:kms:{KEY_RESOURCE_LOCATION}",

Reviewer 2: ```suggestion
              "uri": "{YOUR_GCP_KMS_KEY_RESOURCE_ID_EXCLUDING_VERSION}",
```

>               "auth_principal": "{COCKROACH_CLOUD_SERVICE_ACCOUNT}"

Reviewer 2: ```suggestion
              "auth_principal": "{YOUR_GCP_SERVICE_ACCOUNT_ID}"
```

            }
          },
          {

>             "region": "{COCKROACH_CLOUD_REGION}",

Reviewer 2: ```suggestion
            "region": "{COCKROACHDB_CLOUD_REGION}",
```

            "key_spec": {
              "type": "GCP_CLOUD_KMS",

>               "uri": "{KEY_RESOURCE_LOCATION}",

Reviewer 2: ```suggestion
              "uri": "{YOUR_GCP_KMS_KEY_RESOURCE_ID_EXCLUDING_VERSION}",
```

>               "auth_principal": "{COCKROACH_CLOUD_SERVICE_ACCOUNT}"

Reviewer 2: ```suggestion
              "auth_principal": "{YOUR_GCP_SERVICE_ACCOUNT_ID}"
```

            }
          }
        ]
      }
      ```

    {{site.data.alerts.callout_info}}

>     If you add a new region to a cluster, data at rest in the new region is not automatically protected by CMEK. To enable CMEK for the new region, "rotate" that region to begin using a CMEK key.

Reviewer 2: Let's not include this please, as discussed in Slack.

  --header "Authorization: Bearer ${API_KEY}"
```

## Rotate a CMEK key

A CMEK key can be rotated within your KMS platform or within {{ site.data.products.dedicated }}.

- When you rotate a CMEK key within your KMS platform, a new version of the key is created with new key material. In both GCP KMS and AWS KMS, encryption operations automatically uses the active key version, while decryption operations automatically use the key version that was used to encrypt the data. For this reason, {{ site.data.products.dedicated }} does not need any awareness of rotation operations within your KMS platform.

- When you rotate a CMEK key using the {{ site.data.products.db }} API, you update your cluster's configuration to begin using a different CMEK key for one or more regions. {{ site.data.products.dedicated }} begins using the new key to protect the store key on nodes in that cluster region. In a similar way, after adding a new region to a cluster, you can begin using CMEK to protect data at rest in that region by "rotating" that region from using no CMEK key to using a CMEK key.

> The API to rotate a CMEK key is nearly identical to the API for [activating CMEK on a cluster](#step-4-activate-cmek-for-your-cockroachdb-dedicated-cluster), with one notable exception. When you activate CMEK, you use a `POST` request that includes a CMEK key for each of the cluster's regions. When you rotate a CMEK key, you use a `PUT` request and include a CMEK key for each region whose CMEK key to rotate.

Reviewer 3: Comms around this are somewhat tricky until the next prod release, and i'm not sure how we want to navigate them:

* Right now: `PUT` requests _are not_ allowed on non-`ENABLED` clusters. `PUT` is used for exclusively rotation, not initial-enablement. (I guess PUT-for-add-region is a divergence addressed in other comments).
* After next week: `PUT` requests _are_ allowed for non-`ENABLED` clusters. However, `PUT` for initial-enablement must have all regions.

Author: Discussed offline and we will update the docs when the new behavior is available. cc/ @MichaelTrestman

> The API to rotate a CMEK key is nearly identical to the API for [activating CMEK on a cluster](#step-4-activate-cmek-for-your-cockroachdb-dedicated-cluster), with one notable exception. When you activate CMEK, you use a `POST` request that includes a CMEK key for each of the cluster's regions. When you rotate a CMEK key, you use a `PUT` request and include a CMEK key for each region whose CMEK key to rotate.

Reviewer: ```suggestion
The API to rotate a CMEK key is nearly identical to the API to [activate CMEK on a cluster](#step-4-activate-cmek-for-your-cockroachdb-dedicated-cluster), with one notable exception. When you activate CMEK, you use a `POST` request that includes a CMEK key for each of the cluster's regions. When you rotate a CMEK key, you use a `PUT` request that includes a CMEK key for each region whose CMEK key you intend to rotate.
```
